### ISSUE TITLE
col.str.extract() multiple groups at once

### ISSUE THREAD

**Issue by sm-Fifteen** [pola-rs/polars#3775]
#### Describe your feature request

I often find myself dealing with having to extract features of interest from a string. Something like, let's say, this:

```py
import polars as pl

iso_specs = pl.DataFrame({"iso_code": ["ISO 1000:1992", "ISO 80000-1:2009", "ISO/IEC 18181-1:2022", "ISO/IEC/IEEE 29148:2018", "ISO/IEC 9075-1:2016", "ISO 9660:1988"]})
```

With a quick (and completely unreadable) regex, I can extract various features from those strings:

![image](https://user-images.githubusercontent.com/516999/175115955-ccd3103c-ac63-4635-bd81-890d4133d108.png)

If I need to turn these features into polars columns, however, while that's completely doable, it's certainly not... great.

```py
import polars as pl

UNREADABLE_REGEX = r"(^ISO(?:/[A-Z]+)*)\s(\d+)(?:-(\d+))?(?::(\d{4}))"

iso_specs = pl.DataFrame({"iso_code": ["ISO 1000:1992", "ISO 80000-1:2009", "ISO/IEC 18181-1:2022", "ISO/IEC/IEEE 29148:2018", "ISO/IEC 9075-1:2016", "ISO 9660:1988"]})

iso_specs.select([
    pl.col("iso_code"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 1).alias("authority"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 2).alias("spec_num"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 3).alias("part_num"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 4).alias("revision_year"),
])
```

Given all these columns come from the same regex, they could all be extracted in one go, but I get the impression that the optimizer may not catch that and will just run the regex 4 times. It's also a bit long to write.

Ideally, I would like something like this:

```py
iso_specs.select([
    pl.col("iso_code"),
    pl.col("iso_code").str.extract_groups(UNREADABLE_REGEX, [1,2,3,4]),
])
```

...but then this runs into problems with the names of multicolumn expressions, since `alias()` only works on single columns and `map_alias()` wouldn't be able to make the difference between each.

------

Since #2848 `split_exact()` [has a workaround for the column name issue](https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.internals.expr.string.ExprStringNameSpace.split_exact.html) it returns a struct that can be unnested, though I'm not personally fond of the ergonomics.

```py
csv_data = csv_data.with_columns([
    pl.col("timestamp").str.strptime(pl.Datetime, fmt="%Y-%m-%d %H:%M:%S", exact=True, strict=True),
    
    pl.col("gps_coords").str.split_exact(";", 1).struct.rename_fields(["lat", "long"]),
]).unnest("gps_coords")

# I need to cut the expression in half so I can unnest the columns first
csv_data = csv_data.with_columns([
    pl.col("lat").cast(pl.Float64),
    pl.col("long").cast(pl.Float64),
])
```

EDIT: Alternatively, without unnesting:

```py
df = pl.DataFrame(
    {
		"package": ["polars","pandas","python"],
		"version": ["0.14.18","1.4.4","3.10.7"],
    }
)

df = df.with_columns([
	pl.col("version").str.split_exact(".", n=2).struct.rename_fields(
 		["semver_major","semver_minor","semver_revision"]
 	).alias("semver_struct")
]).with_columns([
	pl.col("semver_struct").struct.field("semver_major"),
	pl.col("semver_struct").struct.field("semver_minor"),
	pl.col("semver_struct").struct.field("semver_revision"),
])
```

**Comment by lizeyan** [comment]
I need it too

**Comment by sm-Fifteen** [comment]
> I need it too. By the way, it could be much more convenient to extract fields by group names rather than indices

I've yet to see a regex using named capture groups in the wild, but I suppose you're right and that supporting them would be desirable as well. If it's understood that such an expression would take a regex and return a struct with all matched groups, [like @alexander-beedie suggested](https://github.com/pola-rs/polars/issues/3859#issuecomment-1171244791), then it wouldn't be much of a problem to extract a mixed list of numbered and named groups, even in Rust.

**Comment by stinodego** [comment]
I will close this in favor of #7990

Hopefully both can be closed soon by #10179 !

**Comment by stinodego** [comment]
Duplicate of #7990
